connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java

```diff
@@ -258,17 +241,12 @@ private Long ensureLeaderConfig(long maxOffset, WorkerCoordinator coordinator) {
         handleLostAssignments(lostAssignments, newSubmissions, completeWorkerAssignment, memberConfigs);
 
         // Do not revoke resources for re-assignment while a delayed rebalance is active
-        // Also we do not revoke in two consecutive rebalances by the same leader
-        canRevoke = delay == 0 && canRevoke;
-
-        // Compute the connectors-and-tasks to be revoked for load balancing without taking into
-        // account the deleted ones.
-        log.debug("Can leader revoke tasks in this assignment? {} (delay: {})", canRevoke, delay);
-        if (canRevoke) {
+        log.debug("Can leader revoke tasks in this assignment? (delay: {})", delay);
+        if (delay == 0) {
+            // Compute the connectors-and-tasks to be revoked for load balancing without taking into
+            // account the deleted ones.
             Map<String, ConnectorsAndTasks> toExplicitlyRevoke =
-                    performTaskRevocation(activeAssignments, currentWorkerAssignment);
-
-            log.debug("Connector and task to revoke assignments: {}", toRevoke);
+                    performTaskRevocation(configured, currentWorkerAssignmentWithoutDuplication);
```

Comment:
change (3): pass `configured` (total tasks we have), and `currentWorkerAssignmentWithoutDuplication` into `performTaskRevocation` method.

Reply:
Can you clarify why this change is necessary? I ran the new `testTaskAssignmentWhenWorkerJoinAfterRevocation` test case with and without it, and although it fails without this change, it looks like that's more due to frail testing logic with the `assertAssignment` method than an actual bug in the rebalancing logic here. If I remove the `assertAssignment` calls but manually check on the distribution of C/T across the cluster after the fifth rebalance, everything is balanced.

I've also produced a test case that fails with this change but succeeds without it:
```java
    @Test
    public void testNewWorkerAndNewTasksInSameRound() {
        doReturn(Collections.EMPTY_MAP).when(assignor).serializeAssignments(assignmentsCapture.capture());

        // Start with 40 tasks
        configState = clusterConfigState(offset, 1, 40);
        when(coordinator.configSnapshot()).thenReturn(configState);

        // Start with three workers
        memberConfigs = memberConfigs(leader, offset, 0, 2);
        expectGeneration();
        assignor.performTaskAssignment(leader, offset, memberConfigs, coordinator, protocolVersion);
        ++rebalanceNum;
        returnedAssignments = assignmentsCapture.getValue();
        assertDelay(0, returnedAssignments);
        expectedMemberConfigs = memberConfigs(leader, offset, returnedAssignments);
        assertNoReassignments(memberConfigs, expectedMemberConfigs);

        applyAssignments(returnedAssignments);
        memberConfigs = memberConfigs(leader, offset, assignments);
        // Add 2 tasks
        configState = clusterConfigState(offset, 1, 42);
        when(coordinator.configSnapshot()).thenReturn(configState);
        // Add a worker
        memberConfigs.put("worker3", new ExtendedWorkerState(leaderUrl, offset, null));
        expectGeneration();
        assignor.performTaskAssignment(leader, offset, memberConfigs, coordinator, protocolVersion);
        ++rebalanceNum;
        returnedAssignments = assignmentsCapture.getValue();
        assertDelay(0, returnedAssignments);
        expectedMemberConfigs = memberConfigs(leader, offset, returnedAssignments);
        assertNoReassignments(memberConfigs, expectedMemberConfigs);

        applyAssignments(returnedAssignments);
        memberConfigs = memberConfigs(leader, offset, assignments);
        // Rebalance once more as a follow-up to task revocation
        expectGeneration();
        assignor.performTaskAssignment(leader, offset, memberConfigs, coordinator, protocolVersion);
        ++rebalanceNum;
        returnedAssignments = assignmentsCapture.getValue();
        assertDelay(0, returnedAssignments);
        expectedMemberConfigs = memberConfigs(leader, offset, returnedAssignments);
        assertNoReassignments(memberConfigs, expectedMemberConfigs);

        applyAssignments(returnedAssignments);
        memberConfigs = memberConfigs(leader, offset, assignments);
        assertBalancedAssignments(memberConfigs);

        verify(coordinator, times(rebalanceNum)).configSnapshot();
        verify(coordinator, times(rebalanceNum)).leaderState(any());
        verify(coordinator, times(2 * rebalanceNum)).generationId();
        verify(coordinator, times(rebalanceNum)).memberId();
        verify(coordinator, times(rebalanceNum)).lastCompletedGenerationId();
    }

    private void assertBalancedAssignments(Map<String, ExtendedWorkerState> existingAssignments) {
        List<Integer> connectorCounts = existingAssignments.values().stream()
                .map(e -> e.assignment().connectors().size())
                .sorted()
                .collect(Collectors.toList());
        List<Integer> taskCounts = existingAssignments.values().stream()
                .map(e -> e.assignment().tasks().size())
                .sorted()
                .collect(Collectors.toList());

        int minConnectors = connectorCounts.get(0);
        int maxConnectors = connectorCounts.get(connectorCounts.size() - 1);

        int minTasks = taskCounts.get(0);
        int maxTasks = taskCounts.get(taskCounts.size() - 1);

        assertTrue(
                "Assignments are imbalanced. The spread of connectors across each worker is: " + connectorCounts,
                maxConnectors - minConnectors <= 1
        );
        assertTrue(
                "Assignments are imbalanced. The spread of tasks across each worker is: " + taskCounts,
                maxTasks - minTasks <= 1
        );
    }
```